presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/RemoveSingleRowDistinctLimit.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.DistinctLimitNode;
+
+import static io.prestosql.sql.planner.optimizations.QueryCardinalityUtil.isScalar;
+import static io.prestosql.sql.planner.plan.Patterns.DistinctLimit.limit;
+import static io.prestosql.sql.planner.plan.Patterns.distinctLimit;
+
+public class RemoveSingleRowDistinctLimit
+        implements Rule<DistinctLimitNode>
+{
+    private static final Pattern<DistinctLimitNode> PATTERN = distinctLimit().with(limit().matching(limit -> limit != 0));
+
+    @Override
+    public Pattern<DistinctLimitNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(DistinctLimitNode node, Captures captures, Context context)
+    {
+        if (isScalar(node.getSource(), context.getLookup())) {
+            return Result.ofPlanNode(node.getSource());
```

Comment:
I think you are losing output symbols here. See `hashSymbol` in `DistinctLimitNode`. Also notice that `DistinctLimitNode::getOutputSymbols`return `distinctSymbols` which might be different than `node.getSource().getOutputSymbols()`.

I wonder why test didn't find that already, so please make sure that there is test coverage for that. Can you please run your test from `TestLogicalPlanner` with coverage or debugging to see if you rule was triggered?

Reply:
Yes but the `hashSymbol` in `DistinctLimitNode` will be added only if we set `optimize_hash_generation` as true and IIRC it will be added in `HashGenerationOptimizer` which will be invoked after this optimizer. So we can safely assume that `hashSymbol` will be empty.

Reply:
Then please verify in rule that `hashSymbol` is empty. Also verify that `DistinctLimitNode::getOutputSymbols` are same as `node.getSource().getOutputSymbols()`.